crates/chain/src/canonical_task.rs

```diff
@@ -0,0 +1,483 @@
+use crate::canonical_iter::{CanonicalReason, ObservedIn};
+use crate::collections::{HashMap, HashSet, VecDeque};
+use crate::tx_graph::{TxAncestors, TxDescendants};
+use crate::{Anchor, CanonicalView, ChainPosition, TxGraph};
+use alloc::boxed::Box;
+use alloc::collections::BTreeSet;
+use alloc::sync::Arc;
+use alloc::vec::Vec;
+use bdk_core::BlockId;
+use bitcoin::{Transaction, Txid};
+
+/// A request for chain data needed during canonicalization.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CanonicalizationRequest {
+    /// Request to check if a block is in the chain.
+    IsBlockInChain {
+        /// The block to check.
+        block: BlockId,
+        /// The chain tip to check against.
+        chain_tip: BlockId,
+    },
+}
+
+/// A response to a canonicalization request.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CanonicalizationResponse {
+    /// Response to IsBlockInChain request.
+    IsBlockInChain(Option<bool>),
```

Comment:
How do we associate responses to their associated requests? I would imagine that some chain sources may not respond in order.